generator/internal/http_annotation_parser.cc

```diff
@@ -0,0 +1,281 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "generator/internal/http_annotation_parser.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/internal/make_status.h"
+#include "google/cloud/log.h"
+#include "absl/strings/match.h"
+
+namespace google {
+namespace cloud {
+namespace generator_internal {
+namespace {
+
+template <typename T>
+struct ParseSuccess {
+  T value;
+  std::size_t end;  // where parsing finished
+};
+template <typename T>
+ParseSuccess<T> MakeParseSuccess(T value, std::size_t end) {
+  return ParseSuccess<T>{std::move(value), end};
+}
+
+template <typename T>
+using ParseResult = StatusOr<ParseSuccess<T>>;
+
+Status ParseError(absl::string_view input, std::size_t offset,
+                  absl::string_view expected,
+                  internal::ErrorInfoBuilder builder) {
+  return internal::InvalidArgumentError(
+      absl::StrCat("error parsing path template, expected", expected,
+                   " at offset ", offset, "\n", input, "\n",
+                   std::string(offset, ' '), "^"),
+      std::move(builder));
+}
+
+ParseResult<PathTemplate::Segments> ParseSegments(absl::string_view input,
+                                                  std::size_t offset);
+
+ParseResult<PathTemplate::Segments> ParsePlainSegments(absl::string_view input,
+                                                       std::size_t offset);
+
+ParseResult<PathTemplate::Segment> ParsePlainSegment(absl::string_view input,
+                                                     std::size_t offset);
+
+ParseResult<PathTemplate::Segment> ParseSegment(absl::string_view input,
+                                                std::size_t offset);
+
+ParseResult<PathTemplate::Segment> ParseVariable(absl::string_view input,
+                                                 std::size_t offset);
+
+ParseResult<PathTemplate::Segments> ParsePlainSegments(absl::string_view input,
+                                                       std::size_t offset) {
+  PathTemplate::Segments segments;
+  while (input.size() != offset) {
```

Comment:
Nit:  I would tend to write these tests the other was around, `offset != input.size()`, given that it is `offset` that is growing and `input.size()` remaining constant.

Reply:
Done.